File: frontend/src/lib/api/preferences.ts
// User preferences API client (unit system, locale, time format).
// Backend: /user/preferences — JSONB on User, values stored server-side.

import type { UserPreferences, UserPreferencesUpdate } from './types';

type AuthFetch = <T>(path: string, options?: RequestInit) => Promise<T>;

export type { UserPreferences, UserPreferencesUpdate };

export async function getPreferences(
  authFetch: AuthFetch,
): Promise<UserPreferences> {
  return authFetch<UserPreferences>('/api/v1/user/preferences');
}

export async function updatePreferences(
  authFetch: AuthFetch,
  payload: UserPreferencesUpdate,
): Promise<UserPreferences> {
  return authFetch<UserPreferences>('/api/v1/user/preferences', {
    method: 'PATCH',
    body: JSON.stringify(payload),
  });
}

File: backend/app/api/preferences.py
"""User UI preferences endpoints (unit system, locale, time format)."""

from fastapi import APIRouter, Depends, HTTPException
from sqlalchemy.ext.asyncio import AsyncSession

from app.database import get_db
from app.models.user import User
from app.schemas.preferences import UserPreferences, UserPreferencesUpdate
from app.services.auth import get_current_user
from app.services.preferences import get_preferences, set_preferences

router = APIRouter()


@router.get("/preferences", response_model=UserPreferences)
async def read_preferences(
    current_user: User = Depends(get_current_user),
):
    """Get the current user's UI preferences (unit system, locale, time format)."""
    return get_preferences(current_user)


@router.patch("/preferences", response_model=UserPreferences)
async def update_preferences(
    payload: UserPreferencesUpdate,
    db: AsyncSession = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    """Update UI preferences (partial update supported)."""
    try:
        return await set_preferences(
            db, current_user, payload.model_dump(exclude_unset=True)
        )
    except ValueError as exc:
        raise HTTPException(status_code=422, detail=str(exc)) from exc
